src/util/prometheus/Prometheus.h

```diff
@@ -0,0 +1,161 @@
+     *
+     * @param name The name of the metric
+     * @param labels The labels of the metric
+     * @param description The description of the metric
+     */
+    virtual GaugeInt&
+    gaugeInt(std::string name, Labels labels, std::optional<std::string> description = std::nullopt) = 0;
+
+    /**
+     * @brief Get a double based gauge metric. It will be created if it doesn't exist
+     *
+     * @param name The name of the metric
+     * @param labels The labels of the metric
+     * @param description The description of the metric
+     */
+    virtual GaugeDouble&
+    gaugeDouble(std::string name, Labels labels, std::optional<std::string> description = std::nullopt) = 0;
+
+    /**
+     * @brief Collect all metrics and return them as a string in Prometheus format
+     *
+     * @return The serialized metrics
+     */
+    virtual std::string
+    collectMetrics() = 0;
+
+    /**
+     * @brief Whether prometheus is enabled
+     *
+     * @return true if prometheus is enabled
+     */
+    bool
+    isEnabled() const
+    {
+        return isEnabled_;
+    }
+
+private:
+    bool isEnabled_;
+};
+
+/**
+ * @brief Implemetation of PrometheusInterface
+ *
+ * @note When prometheus is disabled, all metrics will still counted but collection is disabled
+ */
+class PrometheusImpl : public PrometheusInterface {
+public:
+    using PrometheusInterface::PrometheusInterface;
+
+    CounterInt&
+    counterInt(std::string name, Labels labels, std::optional<std::string> description) override;
+
+    CounterDouble&
+    counterDouble(std::string name, Labels labels, std::optional<std::string> description) override;
+
+    GaugeInt&
+    gaugeInt(std::string name, Labels labels, std::optional<std::string> description) override;
+
+    GaugeDouble&
+    gaugeDouble(std::string name, Labels labels, std::optional<std::string> description) override;
+
+    std::string
+    collectMetrics() override;
+
+private:
+    MetricBase&
+    getMetric(std::string name, Labels labels, std::optional<std::string> description, MetricType type);
+
+    std::unordered_map<std::string, MetricsFamily> metrics_;
+    friend class PrometheusSingleton;
+};
+
+class PrometheusSingleton {
+public:
+    void static init(Config const& config = Config{})
+    {
+        bool const enabled = config.valueOr("prometheus_enabled", true);
+        instance_ = std::make_unique<PrometheusImpl>(enabled);
+    }
+
+    static PrometheusInterface&
+    instance()
+    {
+        assert(instance_);
+        return *instance_;
+    }
+
+    // Be careful with this method because there could be hanging references to counters
+    static void
+    replaceInstance(std::unique_ptr<PrometheusInterface> instance)
+    {
+        instance_ = std::move(instance);
+    }
+
+private:
+    static std::unique_ptr<PrometheusInterface> instance_;
+};
+
+}  // namespace util::prometheus
+
+#define PROMETHEUS util::prometheus::PrometheusSingleton::instance
```

Comment:
I'd prefer to avoid introducing macros. Let's try to make the API closer to `LogService`, wdyt?